### PR TITLE
Fix: Crash on incomplete program

### LINKED ISSUE #2106: Crash in `CheckIsCompilable`
Input:

```dafny
predicate method P(x: bool)
type WE = e: (int, int) | P()
```

Output:

```
crash.dfy(2,27): Error: wrong number of arguments (predicate 'P' expects 1, got 0)
Unhandled exception. System.AggregateException: One or more errors occurred. (Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index'))
 ---> System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.Collections.Generic.List`1.get_Item(Int32 index)
   at Microsoft.Dafny.ExpressionTester.CheckIsCompilable(Expression expr, ICodeContext codeContext) in C:\dafny\Source\Dafny\Util.cs:line 855
   at Microsoft.Dafny.ExpressionTester.CheckIsCompilable(Expression expr, ICodeContext codeContext) in C:\dafny\Source\Dafny\Util.cs:line 985
   at Microsoft.Dafny.ExpressionTester.CheckIsCompilable(Resolver resolver, Expression expr, ICodeContext codeContext) in C:\dafny\Source\Dafny\Util.cs:line 790
   at Microsoft.Dafny.Resolver.ResolveTopLevelDecls_Core(List`1 declarations, Graph`1 datatypeDependencies, Graph`1 codatatypeDependencies, Boolean isAnExport) in C:\dafny\Source\Dafny\Resolver.cs:line 2710
   at Microsoft.Dafny.Resolver.ResolveModuleDefinition(ModuleDefinition m, ModuleSignature sig, Boolean isAnExport) in C:\dafny\Source\Dafny\Resolver.cs:line 1032
   at Microsoft.Dafny.Resolver.ResolveProgram(Program prog) in C:\dafny\Source\Dafny\Resolver.cs:line 507
   at Microsoft.Dafny.Main.Resolve(Program program, ErrorReporter reporter) in C:\dafny\Source\Dafny\DafnyMain.cs:line 162
   at Microsoft.Dafny.Main.ParseCheck(IList`1 files, String programName, ErrorReporter reporter, Program& program) in C:\dafny\Source\Dafny\DafnyMain.cs:line 113
   at Microsoft.Dafny.DafnyDriver.ProcessFiles(IList`1 dafnyFiles, ReadOnlyCollection`1 otherFileNames, ErrorReporter reporter, Boolean lookForSnapshots, String programId) in C:\dafny\Source\DafnyDriver\DafnyDriver.cs:line 249
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at Microsoft.Dafny.DafnyDriver.ThreadMain(String[] args) in C:\dafny\Source\DafnyDriver\DafnyDriver.cs:line 67
   at Microsoft.Dafny.DafnyDriver.<>c__DisplayClass6_0.<Main>b__0() in C:\dafny\Source\DafnyDriver\DafnyDriver.cs:line 47
   at System.Threading.Thread.StartCallback()
```

### PR DESCRIPTION
**Pull request by MikaelMayer**
Fixes #2106 by adding a comparison.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Test/git-issues/git-issue-2106.dfy [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// RUN: %dafny_0 /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+predicate method P(x: bool)
+type WE = e: (int, int) | P()
```
Comment (fabiomadge) [inline Test/git-issues/git-issue-2106.dfy:5]:
```suggestion
type WE = e: (int, int) | P()

```

File: Source/Dafny/Util.cs [hunk 2]
```diff
@@ -858,7 +858,7 @@ public class ExpressionTester {
           }
           // function is okay, so check all NON-ghost arguments
           isCompilable = CheckIsCompilable(callExpr.Receiver, codeContext);
-          for (var i = 0; i < callExpr.Function.Formals.Count; i++) {
+          for (var i = 0; i < callExpr.Function.Formals.Count && i < callExpr.Args.Count; i++) {
             if (!callExpr.Function.Formals[i].IsGhost) {
```
Comment (fabiomadge) [inline Source/Dafny/Util.cs:862]:
```suggestion
          for (var i = 0; i < callExpr.Function.Formals.Count; i++) {
            if (!callExpr.Function.Formals[i].IsGhost  && i < callExpr.Args.Count) {
```

**Review by fabiomadge — APPROVED** [review]
Thanks for taking this on!

**Review by fabiomadge — APPROVED** [review]

File: Source/Dafny/Util.cs (hunk 2 above)
Comment (MikaelMayer) [reply Source/Dafny/Util.cs]: There is no way your suggestion does not imply mine, but for maintenance it's surely better.

File: Source/Dafny/Util.cs [hunk 3]
```diff
@@ -859,7 +859,7 @@ public class ExpressionTester {
           // function is okay, so check all NON-ghost arguments
           isCompilable = CheckIsCompilable(callExpr.Receiver, codeContext);
           for (var i = 0; i < callExpr.Function.Formals.Count; i++) {
-            if (!callExpr.Function.Formals[i].IsGhost) {
+            if (!callExpr.Function.Formals[i].IsGhost  && i < callExpr.Args.Count) {
```
Comment (MikaelMayer) [inline Source/Dafny/Util.cs:862]:
```suggestion
            if (!callExpr.Function.Formals[i].IsGhost && i < callExpr.Args.Count) {
```

**Review by fabiomadge — APPROVED** [review]
Sorry for the extra space.

**Pull request merged**